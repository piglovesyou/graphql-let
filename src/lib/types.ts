export type CommandOpts = {
  cwd: string;
  configPath: string;
};

export type ConfigTypes = {
  generateDir: string;
  config?: Record<string, boolean | string>;
  schema: string;
  documents: string | string[];
  plugins: string[];
};
