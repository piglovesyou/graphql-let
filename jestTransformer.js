module.exports = require('./dist/jestTransformer').default;
