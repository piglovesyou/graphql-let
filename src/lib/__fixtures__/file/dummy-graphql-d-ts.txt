/* a7870383395320c75dc806f677ffa7caad7522bc
 * This file is automatically generated by graphql-let.
 */
