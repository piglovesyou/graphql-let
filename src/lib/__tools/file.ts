import copyfiles from 'copyfiles';
import { promises } from 'fs';
import { join as pathJoin, sep } from 'path';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

const crlf = /\r\n/g;
export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}

export const { readFile: _readFile, writeFile, rename } = promises;

export function readFile(file: string) {
  return _readFile(file, 'utf-8').then(normalizeNewLine);
}

export function cleanup(cwd: string, relPaths: string[]) {
  return Promise.all(relPaths.map((rel) => rimraf(pathJoin(cwd, rel))));
}

export function copyDir(
  baseFullDir: string,
  srcRelDir: string,
  destRelDir: string,
): Promise<void> {
  // baseFullDir: /Users/a/b/c
  // srcRelDir: d/e
  // destRelDir: .d/e
  // It generates /Users/a/b/c/.d/e
  if (srcRelDir === destRelDir) throw new Error('Kidding me?');
  const up = pathJoin(baseFullDir).split(sep).length + 1;
  const [relRoot] = destRelDir.split('/');
  return new Promise<void>((resolve, rejects) => {
    copyfiles(
      [pathJoin(baseFullDir, srcRelDir, '**'), pathJoin(baseFullDir, relRoot)],
      { error: true, up, all: true },
      (err) => {
        if (err) return rejects(err);
        resolve();
      },
    );
  });
}

export type AbsFn = (rel: string) => string;

export async function prepareFixtures(
  baseFullDir: string,
  fixtureSrcRelDir: string,
): Promise<[cwd: string, abs: AbsFn]> {
  if (fixtureSrcRelDir.startsWith('..'))
    throw new Error(
      `It doesn't support backward relative paths like ${fixtureSrcRelDir}`,
    );
  const fixtureDestRelDir = '.' + fixtureSrcRelDir;
  const cwd = pathJoin(baseFullDir, fixtureDestRelDir);
  const abs = (relPath: string) => pathJoin(cwd, relPath);
  await copyDir(baseFullDir, fixtureSrcRelDir, fixtureDestRelDir);
  return [cwd, abs];
}
