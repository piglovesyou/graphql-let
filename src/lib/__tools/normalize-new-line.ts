const crlf = /\r\n/g;

export function normalizeNewLine(str: string) {
  if (crlf.test(str)) return str.replace(crlf, '\n');
  return str;
}
