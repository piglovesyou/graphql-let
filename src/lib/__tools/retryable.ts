function timeout(ms: number) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export default async function retryable(
  fn: () => Promise<void>,
  interval: number,
  maxTimeout: number,
) {
  if (maxTimeout < 0) throw new Error('never');
  const end = Date.now() + maxTimeout;
  let lastError: Error = new Error('never');
  let n = 0;
  while (Date.now() < end) {
    try {
      await fn();
      console.info(`Succeeded on the ${++n} time.`);
      return;
    } catch (e) {
      console.info(`${++n} times tried. Retrying...`);
      lastError = e;
      await timeout(interval);
    }
  }
  throw lastError;
}
