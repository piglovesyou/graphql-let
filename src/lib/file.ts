import { promises as fsPromises } from 'fs';
import { createReadStream, existsSync } from 'fs';
import glob from 'globby';
import _rimraf from 'rimraf';
import { promisify } from 'util';

export const rimraf = promisify(_rimraf);

export const { readFile, writeFile } = fsPromises;
export { readFileSync, statSync } from 'fs';

// Erasing old cache in __generated__ on HMR.
// Otherwise the multiple `declare module "*/x.graphql"` are exposed.
export async function removeByPatterns(cwd: string, ...patterns: string[]) {
  const oldFiles = await glob(patterns, {
    cwd,
    absolute: true,
  });
  await Promise.all(oldFiles.map((f) => rimraf(f)));
}

const leadingStringOfGeneratedContent = '/* ';
const hexHashLength = 40;
export function withHash(sourceHash: string, content: string): string {
  return `${leadingStringOfGeneratedContent}${sourceHash}
 * This file is automatically generated by graphql-let. */

${content}`;
}

export function readHash(filePath: string): Promise<string | null> {
  if (!existsSync(filePath)) {
    return Promise.resolve(null);
  }

  return new Promise((resolve, reject) => {
    const stream = createReadStream(filePath, {
      encoding: 'utf-8',
      highWaterMark: leadingStringOfGeneratedContent.length + hexHashLength,
    });
    stream.on('error', (error) => reject(error));
    stream.on('data', (chunk: string) => {
      const hash = chunk.slice(leadingStringOfGeneratedContent.length);
      if (hash.length !== hexHashLength) return resolve(null);

      resolve(hash);
    });
    stream.read();
  });
}
