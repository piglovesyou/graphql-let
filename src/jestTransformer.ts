import { ProjectConfig } from '@jest/types/build/Config';
import { Transformer } from '@jest/transform';
import { relative as pathRelative } from 'path';
import { readFileSync } from 'fs';
import createExecContext from './lib/exec-context';
import { loadConfigSync } from './lib/config';
import { createPaths } from './lib/paths';
import { createHash } from './lib/hash';

type JestTransformerOptions = {
  subsequentTransformer?: string;
};

function getOption(jestConfig: ProjectConfig): JestTransformerOptions {
  if (!Array.isArray(jestConfig.transform)) return {};
  for (const [, entryPoint, opts] of jestConfig.transform) {
    if (entryPoint.endsWith('graphql-let/jestTransformer.js')) return opts;
  }
  return {};
}

const jestTransformer: Transformer = {
  getCacheKey(fileData, filename, configString) {
    return createHash(fileData + filename + configString + 'graphql-let');
  },
  process(input, filePath, jestConfig, transformOptions) {
    const { rootDir: cwd } = jestConfig;
    const [config, configHash] = loadConfigSync(cwd);
    const execContext = createExecContext(cwd, config, configHash);

    const { tsxFullPath } = createPaths(
      execContext,
      pathRelative(cwd, filePath),
    );
    const tsxContent = readFileSync(tsxFullPath, 'utf-8');

    // Let users customize a subsequent transformer
    const { subsequentTransformer } = getOption(jestConfig);
    if (subsequentTransformer) {
      // eslint-disable-next-line @typescript-eslint/no-var-requires
      return require(subsequentTransformer).process(
        tsxContent,
        tsxFullPath,
        jestConfig,
        transformOptions,
      );
    }

    // "babel-jest" by default
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    const { createTransformer } = require('babel-jest');
    const babelTransformer = createTransformer({ cwd: cwd });
    return babelTransformer.process(
      tsxContent,
      tsxFullPath,
      jestConfig,
      transformOptions,
    );
  },
};

export default jestTransformer;
