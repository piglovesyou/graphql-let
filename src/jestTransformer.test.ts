import { Config } from '@jest/types';
import { join as pathJoin } from 'path';
import compiler from '../test/__tools/compile';
import { AbsFn, prepareFixtures } from '../test/__tools/file';
import jestTransformer from './jestTransformer';

let cwd: string;
let abs: AbsFn;
let jestConfig: Config.ProjectConfig;

describe('graphql-let/jestTransformer', () => {
  beforeAll(async () => {
    [cwd, abs] = await prepareFixtures(__dirname, '__fixtures/jestTransformer');
    jestConfig = { rootDir: cwd } as Config.ProjectConfig;
  });

  test('transforms .graphql', async () => {
    const fileName = 'pages/viewer.graphql';
    const stats = await compiler(cwd, fileName, 'node');
    const { 0: fileData } = stats
      .toJson()
      .modules!.map((m) => m.source)
      .filter(Boolean);

    const fullPath = pathJoin(cwd, fileName);
    const { code: transformedContent } = jestTransformer.process(
      fileData!,
      fullPath,
      jestConfig,
    ) as { code: string };
    expect(removeSourcemapReference(transformedContent)).toMatchSnapshot();
  });
});

function removeSourcemapReference(code: string) {
  // XXX: I couldn't find the better way to suppress sourcemaps
  return code.replace(/\n\/\/# sourceMappingURL=[\s\S]*/, '');
}
