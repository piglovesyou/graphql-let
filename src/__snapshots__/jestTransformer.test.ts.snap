// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql-let/jestTransformer transforms .graphql 1`] = `
"\\"use strict\\";

function _typeof(obj) { \\"@babel/helpers - typeof\\"; if (typeof Symbol === \\"function\\" && typeof Symbol.iterator === \\"symbol\\") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === \\"function\\" && obj.constructor === Symbol && obj !== Symbol.prototype ? \\"symbol\\" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, \\"__esModule\\", {
  value: true
});
exports.useViewerQuery = useViewerQuery;
exports.useViewerLazyQuery = useViewerLazyQuery;
exports.ViewerDocument = void 0;

var Apollo = _interopRequireWildcard(require(\\"@apollo/client\\"));

var _templateObject;

function _getRequireWildcardCache() { if (typeof WeakMap !== \\"function\\") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== \\"object\\" && typeof obj !== \\"function\\") { return { \\"default\\": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj[\\"default\\"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }

var ViewerDocument = (0, Apollo.gql)(_templateObject || (_templateObject = _taggedTemplateLiteral([\\"\\\\n    query Viewer {\\\\n  viewer {\\\\n    id\\\\n    name\\\\n    status\\\\n  }\\\\n}\\\\n    \\"])));
/**
 * __useViewerQuery__
 *
 * To run a query within a React component, call \`useViewerQuery\` and pass it any options that fit your needs.
 * When your component renders, \`useViewerQuery\` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useViewerQuery({
 *   variables: {
 *   },
 * });
 */

exports.ViewerDocument = ViewerDocument;

function useViewerQuery(baseOptions) {
  return Apollo.useQuery(ViewerDocument, baseOptions);
}

function useViewerLazyQuery(baseOptions) {
  return Apollo.useLazyQuery(ViewerDocument, baseOptions);
}"
`;
